### PR TITLE
[builds] WIP... add generic error/warning parser

### PR DESCRIPTION
**Pull request by marcdumais-work**
Signed-off-by: Marc Dumais <marc.dumais@ericsson.com>

### CODE REVIEW

File: packages/builds/src/node/error-parser.spec.ts [hunk 1]
```diff
@@ -127,9 +127,11 @@ describe("error-parser", () => {
             const readStream: NodeJS.ReadableStream = fs.createReadStream(__dirname + logName);
             const promise = parser.parse(gccLinkerErrorMatcher, readStream);
 
+            // TODO: Fix on windows:
+            // "file": "C:\\tmp\\gdb-8.0\\zlib\\inflate.c"
             return expect(promise).to.eventually.deep.equal(
                 [{
-                    "file": "/tmp/gdb-8.0/zlib/inflate.c",
+                    "file": path.resolve("/tmp/gdb-8.0/zlib/inflate.c"),
```
Comment (hexa00) [inline packages/builds/src/node/error-parser.spec.ts]: I think you should use the path class of theia see : core/src/common/path.ts

File: packages/builds/src/node/build-spec-utils.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import * as Stream from "stream";
+import * as Events from "events";
```
Comment (hexa00) [inline packages/builds/src/node/build-spec-utils.ts]: should be lower case stream , events

File: packages/builds/src/node/error-parser.spec.ts [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+describe("error-parser", () => {
+    let parser: IErrorParser;
+    beforeEach(function () {
+        parser = testContainer.get<IErrorParser>(IErrorParser);
+    });
+
+    describe('parse simple gcc error log', () => {
+        it('verify compiler errors and warning are found', () => {
+            const logName: String = '/test-resources/' + 'error.txt';
+            const readStream: NodeJS.ReadableStream = fs.createReadStream(__dirname + logName);
```
Comment (hexa00) [inline packages/builds/src/node/error-parser.spec.ts]: you don't have to specific local variable type like that.. logname = ' ' is OK same for stream

File: packages/builds/src/node/error-parser.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import * as Events from "events";
+import * as path from "path";
```
Comment (hexa00) [inline packages/builds/src/node/error-parser.ts]: * as events

Comment (hexa00) [inline packages/builds/src/node/error-parser.ts]: path should be theia's path class

File: packages/builds/src/node/error-parser.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+export interface IErrorMatcher {
+    // TODO: not all fields are yet used...
+    "name": string,
+    "label": string,
+    /** maps to the build command type? e.g. "typescript", "gcc/g++", ... */
+    "owner": string,
+    /** paths generated are absolute or relative? values: 'absolute' or 'relative' */
+    "fileLocation": string,
+    /** Regexp that matches errors/warnings */
```
Comment (hexa00) [inline packages/builds/src/node/error-parser.ts]: filePath ?  location is within the file ?

File: packages/builds/src/node/error-parser.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+    public parse(errorMatcher: IErrorMatcher, stream: NodeJS.ReadableStream): Promise<IParsedError[]> {
+        return new Promise<IParsedError[]>((resolve, reject) => {
+            try {
+                const errors: IParsedError[] = [];
+
+                // a log chunk has arrived
+                stream.on('data', (chunk: Buffer) => {
+                    this.doParse(chunk, errorMatcher).forEach(entry => {
+                        errors.push(entry);
```
Comment (hexa00) [inline packages/builds/src/node/error-parser.ts]: Should this not be handled by the event callback.. like 
stream.on('data', (chunk: Buffer) => { this.doParse(chunk, errorMatcher) }
and then have a listener on the parser events that pushes to errors ?

Comment (hexa00) [inline packages/builds/src/node/error-parser.ts]: We could also write this as:
stream.on('data', this.doParse.bind(this))

File: packages/builds/src/node/error-parser.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                    });
+                });
+
+                stream.on('error', (err: String) => {
+                    reject(err);
```
Comment (hexa00) [inline packages/builds/src/node/error-parser.ts]: This should also fire an event I guess

File: packages/builds/src/node/error-parser.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+                // finished parsing log, emit "done" event
+                stream.on('end', () => {
+                    this.emit('done', {});
+                    resolve(errors);
+                });
+
+            } catch (err) {
+                reject('Problem reading stream');
+            }
```
Comment (hexa00) [inline packages/builds/src/node/error-parser.ts]: Event here too

File: packages/builds/src/node/error-parser.ts [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+    private doParse(buffer: Buffer, errorMatcher: IErrorMatcher): IParsedError[] {
+        const parsedEntries: IParsedError[] = [];
+
+        const regex: RegExp = new RegExp(errorMatcher.pattern.regexp, 'gm');
+        let match = undefined;
+
+        let fileAndPath = undefined;
+        while (match = regex.exec(buffer.toString())) {
+            // translate path from relative to absolute, if required
```
Comment (hexa00) [inline packages/builds/src/node/error-parser.ts]: That won't be line/line however no ?

Comment (hexa00) [inline packages/builds/src/node/error-parser.ts]: Possibly we could use something like https://nodejs.org/dist/latest-v7.x/docs/api/readline.html ?

File: packages/builds/src/node/error-parser.ts [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                file: fileAndPath,
+                location: match[errorMatcher.pattern.locationGroup],
+                message: match[errorMatcher.pattern.messageGroup],
+                code: match[errorMatcher.pattern.codeGroup]
+            };
+
+            // emit new entry immediately
+            this.emit('error-found', parsedEntry);
+
+            parsedEntries.push(parsedEntry);
```
Comment (hexa00) [inline packages/builds/src/node/error-parser.ts]: I would go with the events only.

File: packages/builds/src/node/inversify.spec-config.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+/* Inversivy config file */
+import { Container /*, interfaces */ } from "inversify";
+import { IErrorParser, ErrorParser } from "./error-parser";
+
+const testContainer = new Container();
+
+testContainer.bind<ErrorParser>(IErrorParser).to(ErrorParser);
+
```
Comment (hexa00) [inline packages/builds/src/node/inversify.spec-config.ts]: You should make builds a module and load it here

Comment (hexa00) [inline packages/builds/src/node/inversify.spec-config.ts]: And actually I think it should be called something else than builds... ? I know it wast the 1st target but now it's more about matching errors and I think that could go in core ?

File: packages/builds/src/node/error-parser.ts (hunk 5 above)
Comment (marcdumais-work) [inline packages/builds/src/node/error-parser.ts]: Sorry that was not very clear. should be better now. BTW name borrowed from vscode

File: packages/builds/src/node/error-parser.ts (hunk 9 above)
Comment (marcdumais-work) [inline packages/builds/src/node/error-parser.ts]: Indeed, we might have received a partial line at the end of the buffer. 
Awesome suggestion about readline. Seems to work nicely.

File: packages/builds/src/node/error-parser.spec.ts (hunk 1 above)
Comment (marcdumais-work) [inline packages/builds/src/node/error-parser.spec.ts]: Thanks - i was not aware we have our own version of Path. After thinking about it a bit, I think it's easier, while in the back-end back-end, to use the more feature-rich node path. When it comes time to send things to the front-end (e.g. error markers), we can convert the paths to Theia's normalized version.

File: packages/builds/src/node/error-parser.ts (hunk 6 above)
Comment (marcdumais-work) [inline packages/builds/src/node/error-parser.ts]: done - using 1st suggestion

File: packages/builds/src/node/build-spec-utils.ts (hunk 2 above)
Comment (marcdumais-work) [inline packages/builds/src/node/build-spec-utils.ts]: done

File: packages/builds/src/node/error-parser.spec.ts (hunk 3 above)
Comment (marcdumais-work) [inline packages/builds/src/node/error-parser.spec.ts]: done

File: packages/builds/src/node/error-parser.ts (hunk 4 above)
Comment (marcdumais-work) [inline packages/builds/src/node/error-parser.ts]: done

File: packages/builds/src/node/error-parser.ts (hunk 7 above)
Comment (marcdumais-work) [inline packages/builds/src/node/error-parser.ts]: done

File: packages/builds/src/node/error-parser.ts (hunk 8 above)
Comment (marcdumais-work) [inline packages/builds/src/node/error-parser.ts]: done

File: packages/builds/src/node/error-parser.ts (hunk 10 above)
Comment (marcdumais-work) [inline packages/builds/src/node/error-parser.ts]: done - now doParse() only emits errors/warnings it finds, and returns nothing.

File: packages/builds/src/node/inversify.spec-config.ts (hunk 11 above)
Comment (marcdumais-work) [inline packages/builds/src/node/inversify.spec-config.ts]: This is, I think, the only outstanding comment. I'll continue to work on this.

**Comment by marcdumais-work** [comment]
This is not being actively worked-on - closing this PR for now.

**Pull request closed without merging**